dbms/src/TiDB/Decode/JsonBinary.cpp

```diff
@@ -64,6 +69,293 @@ inline T decodeNumeric(size_t & cursor, const StringRef & raw_value)
     return res;
 }
 
+template <typename T>
+inline void encodeNumeric(JsonBinary::JsonBinaryWriteBuffer & write_buffer, T value)
+{
+    toLittleEndianInPlace(value);
+    const char * from = reinterpret_cast<const char *>(&value);
+    write_buffer.write(from, sizeof(T));
+}
+
+template <bool doDecode>
+struct NeedDecode
+{
+};
+
+template <>
+struct NeedDecode<true>
+{
+    using Type = String;
+};
+
+template <>
+struct NeedDecode<false>
+{
+    using Type = void;
+};
+
+template <bool doDecode>
+typename NeedDecode<doDecode>::Type DecodeJson(size_t & cursor, const String & raw_value)
+{
+    size_t base = cursor;
+    UInt8 type = raw_value[cursor++];
+    size_t size = 0;
+
+    switch (type) // JSON Root element type
+    {
+    case JsonBinary::TYPE_CODE_OBJECT:
+        cursor += 4;
+        size = decodeNumeric<UInt32>(cursor, raw_value);
+        break;
+    case JsonBinary::TYPE_CODE_ARRAY:
+        cursor += 4;
+        size = decodeNumeric<UInt32>(cursor, raw_value);
+        break;
+    case JsonBinary::TYPE_CODE_LITERAL:
+        size = 1;
+        break;
+    case JsonBinary::TYPE_CODE_INT64:
+    case JsonBinary::TYPE_CODE_UINT64:
+    case JsonBinary::TYPE_CODE_FLOAT64:
+        size = 8;
+        break;
+    case JsonBinary::TYPE_CODE_STRING:
+        size = DecodeVarUInt(cursor, raw_value);
+        size += (cursor - base - 1);
+        break;
+    default:
+        throw Exception(
+            "DecodeJsonBinary: Unknown JSON Element Type:" + std::to_string(type),
+            ErrorCodes::LOGICAL_ERROR);
+    }
+
+    ++size;
+    cursor = base + size;
+    if constexpr (!doDecode)
+        return static_cast<typename NeedDecode<false>::Type>(0);
+    else
+        return static_cast<typename NeedDecode<true>::Type>(raw_value.substr(base, size));
+}
+
+String encodeBase64(const StringRef & str)
+{
+    std::ostringstream oss;
+    Poco::Base64Encoder encoder(oss, Poco::BASE64_NO_PADDING);
+    encoder.rdbuf()->setLineLength(0); /// No newline characters would be added
+    Poco::MemoryInputStream mis(str.data, str.size);
+    Poco::StreamCopier::copyStream(mis, encoder);
+    encoder.close();
+    return oss.str();
+}
+
+bool jsonFinished(std::vector<JsonBinary> & json_binary_vec, bool one)
+{
+    return one && !json_binary_vec.empty();
+}
+
+inline JsonBinary::JsonType getJsonType(const simdjson::dom::element & elem)
+{
+    if (elem.is_object())
+    {
+        return JsonBinary::TYPE_CODE_OBJECT;
+    }
+    else if (elem.is_array())
+    {
+        return JsonBinary::TYPE_CODE_ARRAY;
+    }
+    else if (elem.is_bool() || elem.is_null())
+    {
+        return JsonBinary::TYPE_CODE_LITERAL;
+    }
+    else if (elem.is_number())
+    {
+        return JsonBinary::TYPE_CODE_FLOAT64;
```

Comment:
It is same as tidb
https://github.com/pingcap/tidb/blob/4bbd56b6214226f5114e1c73d75f2851d63936c3/pkg/types/json_binary.go#L516-L518
```
// UseNumber causes the Decoder to unmarshal a number into an interface{} as a
// Number instead of as a float64.
func (dec *Decoder) UseNumber() { dec.d.useNumber = true }
```

Reply:
I think these types are introduced by mysql itself to improve performance(less space, direct operations without cast)(https://datatracker.ietf.org/doc/html/rfc7159). And after cast json to string, both Type_String and Type_DateTime will cast to the same string value, and has no way to keep the type info.

Reply:
My mistake, `double/uint64/int64` should be handled separately. updated.
As for the date type, it seems to be treated as string.